### PR TITLE
[WIP] custom packet serializer implementation

### LINKED ISSUE #406: any support on using msgpack as serializer?

### PR DESCRIPTION
**Pull request by shenjackyuanjie**
yea, here I am again
gogogo!
will fix #406

### CODE REVIEW

File: engineio/Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,7 @@ license = "MIT"
 all-features = true
 
 [dependencies]
-base64 = "0.21.5"
```
Comment (1c3t3a) [inline engineio/Cargo.toml:17]: Please **don't** do unrelated changes as part of your PRs! Why does this needs to be bumped (also to the next major version, which implies a breaking change).

File: engineio/Cargo.toml [hunk 2]
```diff
@@ -29,7 +29,8 @@ async-trait = "0.1.77"
 async-stream = "0.3.5"
 thiserror = "1.0"
 native-tls = "0.2.11"
-url = "2.4.1"
```
Comment (1c3t3a) [inline engineio/Cargo.toml:32]: Same as above? Why is url bumped?

File: engineio/src/header.rs [hunk 3]
```diff
@@ -5,7 +5,6 @@ use http::{
     HeaderValue as HttpHeaderValue,
 };
 use std::collections::HashMap;
-use std::convert::TryFrom;
```
Comment (1c3t3a) [inline engineio/src/header.rs:8]: Why is this necessary? Please don't do unrelated changes

File: engineio/Cargo.toml [hunk 4]
```diff
@@ -29,7 +29,8 @@ async-trait = "0.1.77"
...
+url = "2.5.0"
+rmp = "0.8"
```
Comment (1c3t3a) [inline engineio/Cargo.toml:33]: Can we hide this serializer feature behind a feature flag? I don't think we need this everywhere, but rather have it enabled if necessary. For general work with feature flags, you can take a look here: https://doc.rust-lang.org/cargo/reference/features.html.

File: engineio/src/packet.rs [hunk 5]
```diff
@@ -1,13 +1,36 @@
...
 use serde::{Deserialize, Serialize};
 use std::char;
-use std::convert::TryFrom;
-use std::convert::TryInto;
 use std::fmt::{Display, Formatter, Result as FmtResult, Write};
 use std::ops::Index;
 
 use crate::error::{Error, Result};
+
+pub mod normal;
```
Comment (1c3t3a) [inline engineio/src/packet.rs:11]: I think normal is not an ideal name for this. Rather say "standard"?

File: engineio/src/packet.rs [hunk 6]
```diff
@@ -1,13 +1,36 @@
...
+//     MessagePack,
+// }
+
+// impl Default for PacketSerializer {
+//     fn default() -> Self {
+//         PacketSerializer::Normal
+//     }
+// }
+
+pub trait PacketSerializer{
```
Comment (1c3t3a) [inline engineio/src/packet.rs:29]: The trait is much better (in case the commented code presents an alternative), although I am not sure if I am 100% happy with the definition. I am going to think about it :)

**Review by 1c3t3a — COMMENTED** [review]
Hey @shenjackyuanjie, just a quick first pass :)

File: engineio/Cargo.toml (hunk 4 above)
Comment (shenjackyuanjie) [reply engineio/Cargo.toml]: I have remove this part of code and left only custom serializer support

**Comment by shenjackyuanjie** [comment]
to be clear, the code is not done, and there is some leftover need to be cleanup

**Comment by shenjackyuanjie** [comment]
bad news, I found my implment should be at socketio packet ……
well, close and anther PR will be send
